support more architectures, move to separate folder structure

whysthatso opened the pull request:
and another one. this one might be a bit tricky, as it introduces filesystem changes, like described under Assumptions in the readme.

i'm leaving versioned binaries under /opt/ and link the version defined in the playbook into `/usr/local/bin` which i believe is the canonical location for binaries installed like this. 

in a way, it's a breaking change, cause the role now does not care about preexisting installations of either restic or autorestic under `/usr/bin`. but to be honest i am not sure where the line is on how much a role should care about previous state on a box.

other details:
- i've introduced `autorestic_architecture` as an override. i got a use case in which one of my machines returns `mips` for `uname -m` but is actually a `mipsle`, so i had to somehow circumvent that and make it obvious.
- i've listed the precompiled binary locations under dependencies, because that is what they are
- i removed the mode setting of the downloaded files to 0755 because i don't think this is necessary
- i dabbled with unarchive but that module can't rename extracted files, and does not play nicely with bzip2 so i reverted that to your approach. the versioned files however give us at least a chance to have idempotency with the `creates` directive.

hope this is clear enough

whysthatso committed:
support more architectures, move to separate folder structure

whysthatso committed:
D'Oh

dbrennand commented on README.md:
```diff
@@ -11,10 +11,21 @@ Ansible role to configure backups using [autorestic](https://autorestic.vercel.a
 
+## Assumptions
+
+Places versioned executables into `/opt/autorestic/bin` and `/opt/restic/bin` links these into `/usr/local/bin`.
```
Maybe there should be two role variables for this? Something like `autorestic_install_directory` and `autorestic_restic_install_directory`? Where the defaults are `/usr/local/bin` but can obviously be overridden to whatever the user desires, this would work for your use-case right?


where the user can specify where to install the autorestic and restic binaries to?

This way, we can set the default to `/usr/local/bin`

Something like `autorestic_install_directory` maybe?

whysthatso replied:
how about we define
autorestic_install_dir: /opt/autorestic
autorestic_restic_install_dir: /opt/restic
autorestic_binary_dir: /usr/local/bin
autorestic_restic_binary_dir: /usr/local/bin

or is that too much variety?
come to think of it, naming it is not so trivial it seems. is the install dir the final location of a binary/link to it? that might be a bit overthought.

dbrennand replied:
Can we include the `/bin` in the variable defaults like:

```yaml
autorestic_install_dir: /opt/autorestic/bin
autorestic_restic_install_dir: /opt/restic/bin
```

I see in the actions you've appended `/bin` to the var path. I think this could become confusing to some users who might expect the binary to be exactly in the directory they've specified, only to find it's in that dir + `/bin`.

I think these two variables might be a little too much:

```yaml
autorestic_binary_dir: /usr/local/bin
autorestic_restic_binary_dir: /usr/local/bin
```

`/usr/local/bin` is pretty much always in the `$PATH` so I think it's OK to set this to always be `/usr/local/bin`? What do you think? Can you think of a use-case where this is might be different?

whysthatso replied:
oh i missed this one. done now.

dbrennand commented on tasks/autorestic.yml:
```diff
@@ -1,34 +1,61 @@
+    mode: 0755
+    owner: root
+
 - name: "Download autorestic binary v{{ autorestic_version }} for aarch64"
```
Removing the file permissions causes the Ansible-Lint warning. I think we should just keep the `mode` as it was. Or if you think the permissions are wrong, do you think they should be different? :)

whysthatso replied:
0644 should be enough for the downloaded compressed files, as they do not need to be executable. i readd that.

dbrennand commented on README.md:
```diff
@@ -61,7 +72,22 @@ Whether or not to create an autorestic crontab entry to trigger automated backup
+[restic/restic](https://github.com/restic/restic/releases/)
+[cupcakearmy/autorestic](https://github.com/cupcakearmy/autorestic/releases/)
+
+Currently supporte Linux binaries exist for:
```
Typo here.

dbrennand commented on tasks/autorestic.yml:
```diff
@@ -1,34 +1,61 @@
     mode: +x
+
+- name: Create symbolic link to the correct version
+  file:
```
Please use FQDN for the module action🙂

dbrennand commented on tasks/autorestic.yml:
```diff
@@ -1,34 +1,61 @@
+  ansible.builtin.get_url:
+    url: "https://github.com/cupcakearmy/autorestic/releases/download/v{{ autorestic_version }}/autorestic_{{ autorestic_version }}_linux_{{ item }}.bz2"
+    dest: /tmp/autorestic.bz2
+  when: ansible_architecture == item
+  with_items:
+    - mips
+    - mipsle
+    - mips64
+    - mips64le
+    - ppc64le
+    - s390x
+    - 386
```
This is pretty cool :) - So it will run the action only for the matching item in `ansible_architecture`?

I usually see `with_items` for looping over each item, but this is slightly different? 🙂

whysthatso replied:
the difference really is the `when` condition. as i understand it, `with_items` loops the whole task, so the condition is tested on every single loop run, hence the possibility to use it for our case.

dbrennand reviewed (changes requested):
Left a couple of comments. Let me know what you think 🙂 

We should probably merge #4 first, and include those changes in this PR as I've fixed some of the warnings here in that PR.

whysthatso committed:
fix linting complaints, typo fix

whysthatso committed:
make install and binary directories configurable for both restic and autorestic

whysthatso committed:
all hail the linter god

whysthatso committed:
add missing defaults

whysthatso committed:
Merge branch 'main' into support-more-architectures

whysthatso committed:
add autorestic version that supports multiple architectures

whysthatso committed:
try different syntax

dbrennand commented on tasks/main.yml:
```diff
@@ -7,11 +7,11 @@
 - name: Import removal tasks
   ansible.builtin.import_tasks: removal.yml
-  when: autorestic_state == 'absent'
+  when: autorestic_state is "absent"
```
@whysthatso - Did you have some issues with the syntax here? The molecule tests seemed to be passing OK when merging #4

whysthatso replied:
now it's good. there was a silly directory mistake, i've gotten rid of the intermediary bin. now it's all good.
also, will branch off of main in the future...

whysthatso committed:
syntax fix

whysthatso committed:
simplify directory structure

whysthatso committed:
follow default directory structure assumptions for package installation

whysthatso commented:
anything else amiss?

dbrennand commented on defaults/main.yml:
```diff
@@ -21,4 +21,6 @@ autorestic_config: |-
 autorestic_check: false
 autorestic_cron: false
+autorestic_install_dir: /opt/autorestic
+autorestic_restic_install_dir: /opt/restic
```
Can we make these `/opt/autorestic/bin` and `opt/restic/bin` please?

dbrennand commented on tasks/autorestic.yml:
```diff
@@ -1,34 +1,69 @@
+- name: Ensure autorestic directory exist
+  ansible.builtin.file:
+    state: directory
+    path: "{{ autorestic_install_dir }}/bin"
```
Can we remove the `/bin` from here, and have the absolute path in `autorestic_install_dir`?

dbrennand commented on README.md:
```diff
@@ -69,7 +80,22 @@ Whether or not to remove autorestic, restic, configuration and crontab entry. Se
+[restic/restic](https://github.com/restic/restic/releases/)
+[cupcakearmy/autorestic](https://github.com/cupcakearmy/autorestic/releases/)
+
+Currently supporte Linux binaries for:
```
Currently supported Linux binaries:

dbrennand commented on tasks/restic.yml:
```diff
@@ -1,34 +1,69 @@
+- name: Ensure restic directory exist
+  ansible.builtin.file:
+    state: directory
+    path: "{{ autorestic_restic_install_dir }}/bin"
```
Can we remove the `/bin` from here, and have the absolute path in `autorestic_restic_install_dir`?

dbrennand commented on tasks/restic.yml:
```diff
@@ -1,34 +1,69 @@
 ---
 # restic tasks file for autorestic
+- name: Ensure restic directory exist
```
Ensure restic install directory exists

dbrennand commented on tasks/autorestic.yml:
```diff
@@ -1,34 +1,69 @@
 ---
 # autorestic tasks file for autorestic
+- name: Ensure autorestic directory exist
```
Ensure autorestic install directory exists

dbrennand reviewed (changes requested):
Some minor changes suggested.

Two things about this PR:

1. I'm not sure about the hardcoding of the `owner: root` on several tasks, specifically the download tasks. Do you have a specific use-case where you needed the downloaded file in `/tmp` to be owned by root?

2. From what I understand, these changes would mean that the old versions of the binaries would be kept in `/opt/autorestic|restic/bin`, do I understand that correctly? Shouldn't the old version just be overwritten (which was the old behaviour).

dbrennand committed:
[skip ci]refactor: append `/bin` to default install dir

dbrennand committed:
[skip ci]docs(refactor): modify assumptions

Include example for `autorestic_architecture` and include missing `*_directory` vars.

Simplify dependencies wording.

dbrennand committed:
[skip ci]refactor: move `*_directory` vars up

dbrennand committed:
[skip ci]refactor: task names for install directory creation

dbrennand committed:
refactor: allow `autorestic_install_directory` and `autorestic_restic_install_directory` to be configurable

Allow configurable `owner`, `group` and `mode`

dbrennand commented:
Hey @whysthatso - I made a couple of changes. I made it so `autorestic_install_directory` and `autorestic_restic_install_directory` attributes such as `owner`, `group` and `mode` are configurable but keeps the defaults as before.

Let me know what you think 🙂 - Also one other thing, is overriding the `ansible_architecture` safe? What happens if a user runs this role and then runs a different role after which relies on `ansible_architecture`, but we've modified it, couldn't this cause issues?

Maybe we should just use `autorestic_architecture` only and not override `ansible_architecture`. 🤔 Thoughts?

Thanks 👍🏻

dbrennand committed:
fix: remove install directories too

dbrennand committed:
refactor: add task to reset `ansible_architecture`

dbrennand commented:
@whysthatso - To address concern above, I've added an additional task which gathers facts again to reset `ansible_architecture` when `autorestic_architecture` is defined.

dbrennand commented:
@whysthatso Thank you for your contribution! 👍🏻 😃

whysthatso commented:
@dbrennand thank you so much for finishing this up, and sorry for having abandoned it. life got in the way once again...

i think you found good solutions. regarding leaving around old binaries, and then linking against the version du jour, that's just an attitude :D i think if it is considered that a user wants to downgrade and can easily get older versions from a remote source, it's not a problem at all to overwrite them.
